### PR TITLE
[Merged by Bors] - feat: port Deprecated.Submonoid

### PR DESCRIPTION
**Pull request by mo271**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by mo271** [comment]
https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/.60to_additive.60.20with.20.60inductive.60

**Comment by mo271** [comment]
see mathlib4#1783

File: Mathlib/Deprecated/Submonoid.lean [hunk 1]
```diff
@@ -0,0 +1,409 @@
...
+
+
+-- open BigOperators -- Porting note: commented out locale
+
+variable {M : Type _} [Monoid M] {s : Set M}
+
+variable {A : Type _} [AddMonoid A] {t : Set A}
+
+/-- `s` is an additive submonoid: a set containing 0 and closed under addition.
+Note that this structure is deprecated, and the bundled variant `add_submonoid A` should be
```
Comment (jcommelin) [inline Mathlib/Deprecated/Submonoid.lean:44]:
```suggestion
Note that this structure is deprecated, and the bundled variant `AddSubmonoid A` should be
```

File: Mathlib/Deprecated/Submonoid.lean [hunk 2]
```diff
@@ -0,0 +1,409 @@
...
+preferred. -/
+structure IsAddSubmonoid (s : Set A) : Prop where
+  /-- The proposition that s contains 0. -/
+  zero_mem : (0 : A) ∈ s
+  /-- The proposition that s is closed under addition. -/
+  add_mem {a b} : a ∈ s → b ∈ s → a + b ∈ s
+#align is_add_submonoid IsAddSubmonoid
+
+/-- `s` is a submonoid: a set containing 1 and closed under multiplication.
+Note that this structure is deprecated, and the bundled variant `submonoid M` should be
```
Comment (jcommelin) [inline Mathlib/Deprecated/Submonoid.lean:54]:
```suggestion
Note that this structure is deprecated, and the bundled variant `Submonoid M` should be
```

File: Mathlib/Deprecated/Submonoid.lean [hunk 3]
```diff
@@ -0,0 +1,409 @@
...
+      "The sum of a list of elements of an `AddSubmonoid` is an element of the `AddSubmonoid`."]
+theorem list_prod_mem (hs : IsSubmonoid s) : ∀ {l : List M}, (∀ x ∈ l, x ∈ s) → l.prod ∈ s
+  | [], _ => hs.one_mem
+  | a :: l, h =>
+    suffices a * l.prod ∈ s by simpa
+    have : a ∈ s ∧ ∀ x ∈ l, x ∈ s := by simpa using h
+    hs.mul_mem this.1 (list_prod_mem hs this.2)
+#align is_submonoid.list_prod_mem IsSubmonoid.list_prod_mem
+
+/-- The product of a multiset of elements of a submonoid of a `comm_monoid` is an element of
```
Comment (jcommelin) [inline Mathlib/Deprecated/Submonoid.lean:231]:
```suggestion
/-- The product of a multiset of elements of a submonoid of a `CommMonoid` is an element of
```

File: Mathlib/Deprecated/Submonoid.lean [hunk 4]
```diff
@@ -0,0 +1,409 @@
...
+@[to_additive "The image of an `AddMonoid` hom is an `AddSubmonoid` of the codomain."]
+theorem Range.is_submonoid {γ : Type _} [Monoid γ] {f : M → γ} (hf : IsMonoidHom f) :
+    IsSubmonoid (Set.range f) := by
+  rw [← Set.image_univ]
+  exact Univ.IsSubmonoid.image hf
+#align range.is_submonoid Range.is_submonoid
+
+/-- Submonoids are closed under natural powers. -/
+@[to_additive
+      "An `add_submonoid` is closed under multiplication by naturals."]
```
Comment (jcommelin) [inline Mathlib/Deprecated/Submonoid.lean:198]:
```suggestion
      "An `AddSubmonoid` is closed under multiplication by naturals."]
```

File: Mathlib/Deprecated/Submonoid.lean [hunk 5]
```diff
@@ -0,0 +1,409 @@
...
+/-- An element of a monoid is in the set of that element's natural number powers. -/
+@[to_additive
+      "An element of an `AddMonoid` is in the set of that element's natural number multiples."]
+theorem powers.self_mem {x : M} : x ∈ powers x :=
+  ⟨1, pow_one _⟩
+#align powers.self_mem powers.self_mem
+
+/-- The set of natural number powers of an element of a monoid is closed under multiplication. -/
+@[to_additive
+      "The set of natural number multiples of an element of an `add_monoid` is closed under
```
Comment (jcommelin) [inline Mathlib/Deprecated/Submonoid.lean:146]:
```suggestion
      "The set of natural number multiples of an element of an `AddMonoid` is closed under
```

File: Mathlib/Deprecated/Submonoid.lean [hunk 6]
```diff
@@ -0,0 +1,409 @@
...
+## Main definitions
+
+`isAddSubmonoid (S : Set M)` : the predicate that `S` is the underlying subset of an additive
+submonoid of `M`. The bundled variant `AddSubmonoid M` should be used in preference to this.
+
+`isSubmonoid (S : Set M)` : the predicate that `S` is the underlying subset of a submonoid
+of `M`. The bundled variant `Submonoid M` should be used in preference to this.
+
+## Tags
+Submonoid, Submonoids, isSubmonoid
```
Comment (jcommelin) [inline Mathlib/Deprecated/Submonoid.lean:33]:
```suggestion
Submonoid, Submonoids, IsSubmonoid
```

File: Mathlib/Deprecated/Submonoid.lean [hunk 7]
```diff
@@ -0,0 +1,409 @@
...
+
+This file defines unbundled multiplicative and additive submonoids. Instead of using this file,
+please use `Submonoid G` and `AddSubmonoid A`, defined in `GroupTheory.Submonoid.Basic`.
+
+## Main definitions
+
+`isAddSubmonoid (S : Set M)` : the predicate that `S` is the underlying subset of an additive
+submonoid of `M`. The bundled variant `AddSubmonoid M` should be used in preference to this.
+
+`isSubmonoid (S : Set M)` : the predicate that `S` is the underlying subset of a submonoid
```
Comment (jcommelin) [inline Mathlib/Deprecated/Submonoid.lean:29]:
```suggestion
`IsAddSubmonoid (S : Set M)` : the predicate that `S` is the underlying subset of an additive
submonoid of `M`. The bundled variant `AddSubmonoid M` should be used in preference to this.

`IsSubmonoid (S : Set M)` : the predicate that `S` is the underlying subset of a submonoid
```

File: Mathlib/Deprecated/Submonoid.lean [hunk 8]
```diff
@@ -0,0 +1,409 @@
...
+    fun ⟨y, hy, z, hz, hyzx⟩ =>
+    hyzx ▸
+      (closure.IsSubmonoid _).mul_mem (closure_mono (Set.subset_union_left _ _) hy)
+        (closure_mono (Set.subset_union_right _ _) hz)⟩
+#align monoid.mem_closure_union_iff Monoid.mem_closure_union_iff
+
+end Monoid
+
+/-- Create a bundled submonoid from a set `s` and `[isSubmonoid s]`. -/
+@[to_additive "Create a bundled additive submonoid from a set `s` and `[isAddSubmonoid s]`."]
```
Comment (jcommelin) [inline Mathlib/Deprecated/Submonoid.lean:401]:
```suggestion
/-- Create a bundled submonoid from a set `s` and `[IsSubmonoid s]`. -/
@[to_additive "Create a bundled additive submonoid from a set `s` and `[IsAddSubmonoid s]`."]
```

**Review by jcommelin — COMMENTED** [review]
Thanks!

bors d+

**Comment by mo271** [comment]
thanks
bors r+

**Pull request closed without merging**